### DISCUSSION TITLE
Audio CD playback

### DISCUSSION THREAD

**Discussion by vale-max** [rern/rAudio#302] [General]
Hi Rern,
great job with rAudio, specially on FLAC/cue handling.
Have you considered adding playback for CD audio from an USB CD reader?

**Comment by vale-max** [comment]
@rern 
as I had some time I added this feature, in a very embryo and imperfect way, but it works.
When I insert a cd in the reader the system automagically looks for info on cddb and for some cover, then adds the tracks to mpd playlist and start playing them.
If the eject button is pressed, the cdda tracks are removed from the playlist and finally the CD is ejected (I noticed that ejecting while MPD is playing crashes it)

The modifications I did are:
changed /etc/conf.d/devmon (modified to execute "mpc-cd.sh add" at cd insertion)
added /etc/udev/rules.d/60-cdrom_id.rules (took from lib and modified to execute  "mpc-cd remove" when eject is pressed on cd unit )

added /srv/http/bash/mpc-cd.sh to recognize the disc, query cddb, find a cover with glyr and add/remove the tracks (originally from [Mathieu Requillart's blog](https://mathieu-requillart.medium.com/my-ultimate-guide-to-the-raspberry-pi-audio-server-i-wanted-audio-cd-f985e8bd832c) )
modified /srv/http/bash/status.sh to handle cdda files during play

created dir /srv/http/data/cd_current and added a  default.png
created dir /srv/http/data/cd_cache

additional packages needed
abcde
glyr

Todo: the playlist view doesn't like the cdda files at all

[raudio_cdplay.tar.gz](https://github.com/rern/rAudio-1/files/6418713/raudio_cdplay.tar.gz)

**Comment by vale-max** [comment]
I also (poorly) fixed mpdplaylist.php in order to avoid it exiting with error if a cdda element is present in the current playlist.

@rern I'm sorry for the poor quality of this implementation but I'm not a programmer/developer by any length, just wanted to share what I did in case anyone else wants to play its audio CDs with rAudio-1.

[mpdplaylist.php.gz](https://github.com/rern/rAudio-1/files/6462336/mpdplaylist.php.gz)

**Reply by rern** [reply]
@vale-max,

Thank you for your codes. It guided me to the right direction. It has turned out we need only `cd-discid` package.

Next update will have the audio CD support:
- Re-configure on power on/off
- Add tracks individually to Playlist on insert / power on with inserted CD
- Complete with coverart(from `last.fm`) and metadata: artist title album duration (from `gnudb.org`)
- Behave virtually like normal files in Playback
- Can be rearranged order / remove individually in Playlist
- Remove all CD tracks on eject / power off

Please try `UPDATE` branch:
- Remove all of your codes
- Settings > Addons > rAudio 1 - long-press `Update` button
- OK for `UPDATE` branch
- Try:
  - CD drive - power on/off
  - CD drive - connect/disconnect
  - Disc - insert/eject
  - rAudio - reboot/power on with inserted CD

**Comment by R101G** [comment]
@rern 

I like this and gave it a try, but power voltage dropped too low with a drive connected, and I ended up with an un-bootable SD card. I noticed that the alternative setting for no cover art does not stick in this update.

**Reply by vale-max** [reply]
It works ok for me on a rpi4, it worked on a rpi2 but with the drive connected to a powered USB hub.

**Comment by vale-max** [comment]
Hi @rern ,
I built a new image as I had fiddled a lot with the other installation so I wanted to try your implementation in a clean env.
I used the rOS installation to create an RPi4 aarch64 image using update branch directly, and it works. When I insert the CD the tracks are added and removed as I eject it, with no side effects.
Thanks a lot!!

Can I suggest  a couple of improvements?
 
 - Autoplay as a CD is inserted (maybe selectable somewhere in the UI), so even a kid can listen to music without having to open any interface (yes, I have kids :) )

 - An "Audio CD" button in the library view to be able to insert CD tracks in the playlist if the CD is already in and the tracks have been removed previously.
